src/main/java/bio/terra/common/FutureUtils.java

```diff
@@ -1,87 +1,90 @@
-      List<T> returnList =
-          stream
-              .map(
-                  f -> {
-                    try {
-                      // If a failure was found, all subsequent tasks should be canceled
-                      if (foundFailure.get().isPresent()) {
-                        f.cancel(true);
-                      } else if (maxThreadWait.isPresent()) {
-                        return f.get(maxThreadWait.get().toMillis(), TimeUnit.MILLISECONDS);
-                      } else {
-                        return f.get();
-                      }
-                    } catch (TimeoutException e) {
-                      if (!foundFailure.get().isPresent()) {
-                        foundFailure.set(Optional.of(new ApiException("Thread timed out", e)));
-                        // Cancel the thread
-                        f.cancel(true);
-                      }
-                    } catch (InterruptedException e) {
-                      if (!foundFailure.get().isPresent()) {
-                        foundFailure.set(
-                            Optional.of(new ApiException("Thread was interrupted", e)));
-                      }
-                    } catch (ExecutionException e) {
-                      if (!foundFailure.get().isPresent()) {
-                        foundFailure.set(
-                            Optional.of(new ApiException("Error executing thread", e)));
-                      }
+  @VisibleForTesting
+  static <T> List<@NotNull T> waitFor(final List<Future<T>> futures, final Duration maxThreadWait) {
+    final AtomicReference<ErrorReportException> foundFailure = new AtomicReference<>();
+    List<T> returnList =
+        futures.stream()
+            .map(
+                f -> {
+                  try {
+                    // If a failure was found, all subsequent tasks should be canceled
+                    if (foundFailure.get() != null) {
+                      f.cancel(true);
+                    } else if (maxThreadWait != null) {
+                      return f.get(maxThreadWait.toMillis(), TimeUnit.MILLISECONDS);
+                    } else {
+                      return f.get();
+                    }
+                  } catch (TimeoutException e) {
+                    foundFailure.compareAndSet(
+                        null, new ServiceUnavailableException("Thread timed out", e));
+                    f.cancel(true); // Do we need this?
+                  } catch (InterruptedException e) {
+                    foundFailure.compareAndSet(null, new ApiException("Thread was interrupted", e));
+                    f.cancel(true); // Do we need this?
+                  } catch (ExecutionException e) {
+                    if (e.getCause() instanceof ErrorReportException ere) {
+                      // We do not wrap an ErrorReportException cause to preserve its HTTP status.
+                      foundFailure.compareAndSet(null, ere);
+                    } else {
```

Comment:
Flagging for reviewers: if there's some existing code that I can reuse here for this exception handling, I'm all ears!

Reply:
Unfortunately, I don't think so.  This is pretty specific to this case